pygfx/utils/transform.py

```diff
@@ -0,0 +1,97 @@
+import numpy as np
+import pylinalg as pla
+from datetime import datetime
+
+
+class AffineTransform:
+    """An affine tranformation"""
+
+    def __init__(
+        self, matrix=None, /, *, position=None, rotation=None, scale=None
+    ) -> None:
+        self.last_updated = datetime.now()
+
+        if matrix is None:
+            matrix = np.eye(4, dtype=float)
+
+        self.matrix = np.asarray(matrix)
+        self._cache_time = datetime(1900, 1, 1)  # some date in the past
```

Comment:
`time.time` will return a time in seconds not nanoseconds, which I think is too coarse for us. The other advantage of using a `datetime` is that it is compatible with numpy's `datetime64`, which I think will come in handy when I implement the caching of world matrices.

Reply:
Even better: `time.perf_counter()`, since it does not need to be an actual timestamp either.

Reply:
`time.perf_counter()` could work, but it will break down if we want to allow serializing/storing/restoring the renderer state I think.

Reply:
But when you serialize, I freeze all the data and time stops being relevant I would think.

If I load it the transform, couldn't I just reset this value?

Do you perhaps have a use case in mind where this would be a useful property that I'm not seeing?

Reply:
> But when you serialize, I freeze all the data and time stops being relevant I would think.

It depends on which parts we store. For `AffineTransform` it would be enough to only serialize/store `self.matrix`. Position, rotation, and scale are simply cached values because matrix decomposition is expensive so it is best avoided/deferred. All other attributes (timestamps) supplement this caching mechanism and allow updating the cache lazily (just before the actual read).

If we only keep `self.matrix` then we effectively discard the cache and will have to rebuild it during load or when resuming the application. This can be okay, but if things are time-critical (or the scene is large) then it is much faster to store/load the cache as well.

The problem with doing this while using `time.perf_counter()` is that it doesn't guarantee a common reference point. So if I restore the cache then `self._cache_time` will use the (unknown) reference from the previous run; however, `self.last_modified` may change now use a new reference point. This means that the behavior of `self.last_modified > self._cache_time` is now undefined and we no longer know when our cache is invalid.

---

The other use-case for using a full timestamp for `self.last_modified` is its compatibility with numpy. This allows us to store `AffineTransform`s as a structure of arrays (compared to an array of structs) which enables cache-optimized vectorization when working on groups of AffineTransforms (e.g. when caching the WorldMatricies).

Reply:
What kind of serialization are you thinking of? I would expect that decomposing the matrix would generally be a small cost compared to the serialization itself, and serialization is not something that you'd do every frame (but maybe you have a use-case in mind that I am overlooking).

I did a quick-n-dirty [benchmark](https://gist.github.com/almarklein/20640253fe54fc35d4a4c6dfcc1ad625) and it looks like `datetime` is indeed a bit more precise than `time.time()`. However, `perf_counter_ns` is even more precise (it never produced the same value) and is also 3x faster than `datetime`. (Same for `perf_counter`.)

I'm also a bit reluctant of using a somewhat complex datatype for this task. E.g. the benchmark I did also revealed that converting a `datetime` to a numpy `datetime64` is relatively costly. It's easy to introduce stuff like this accidentally.

To summarize, I would suggest using `perf_counter` unless there is a really good reason to require absolute timestamps.

Reply:
Interesting. The benchmark creates a new object each iteration, so you are mostly measuring the creation time of `datetime` vs the creation time of `perf_counter`. What's critical for us is the comparison of two elements, but that is still about 20% faster for `perf_counter_ns` vs `datetime`. 

<details>

```python
from datetime import datetime

import numpy as np

import time

n = 100000


mark1 = datetime.now()
mark2 = datetime.now()
t0 = time.perf_counter()
count = 0
for i in range(n):
    if mark1 < mark2:
        mark1 = mark2
        count += 1
print("Datetime:                   ", f"{time.perf_counter() - t0:.4f}", count)


mark1 = np.datetime64(datetime.now())
mark2 = np.datetime64(datetime.now())
t0 = time.perf_counter()
count = 0
for i in range(n):
    if mark1 < mark2:
        mark1 = mark2
        count += 1
print("Numpy conversion + datetime:", f"{time.perf_counter() - t0:.4f}", count)

mark1 = time.time()
mark2 = time.time()
t0 = time.perf_counter()
count = 0
for i in range(n):
    if mark1 < mark2:
        mark1 = mark2
        count += 1
print("Perf counter:               ", f"{time.perf_counter() - t0:.4f}", count)

mark1 = time.perf_counter_ns()
mark2 = time.perf_counter_ns()
t0 = time.perf_counter()
count = 0
for i in range(n):
    if mark1 < mark2:
        mark1 = mark2
        count += 1
print("Perf counter ns:            ", f"{time.perf_counter() - t0:.4f}", count)
```

</details>

Theoretically, this shouldn't be the case (both are simple integers internally), but since it is this way in Python let's go with `perf_counter_ns`.

Reply:
Looking at [the source for datetime objects](https://github.com/python/cpython/blob/main/Lib/datetime.py) comparison is implemented with `isinstance` which is notably slow and the comparison of tuples of year/month/day etc.

That has to be slower when contrasted to plain ints as returned by perf_counter_ns.

Makes sense!

Reply:
> Looking at [the source for datetime objects](https://github.com/python/cpython/blob/main/Lib/datetime.py) comparison is implemented with isinstance which is notably slow and the comparison of tuples of year/month/day etc.

Jeesus. Thanks @almarklein for being adamant about this. Let's definitely not use `datetime` in that case >.<